### PR TITLE
Move more things from the view layer to the service model 

### PR DESCRIPTION
**Pull request by quis**
We do a lot of logic inside the view methods.

It makes it cleaner to break this logic up into functions. But this would mean passing around variables from one function to another. Putting these methods onto a class (the service model) means that
there’s a place to store this data (rather than having to pass it around a lot).

Making this code more manageable is important so that when we have templates and folders it’s easy to encapsulate the logic around combining the two.

# Before 

img_8604

# After 

img_8605

***

Notes for reviewing:

- There’s a lot of refactoring and moving things around to make this work possible. I‘ve tried to break it into sensible commits, and as such is best reviewed one commit at a time.
- The two view methods I really focused on were `choose_template` and `service_settings`. It would have been too much work to refactor the whole app in one pull request, but hopefully this shows:
    - The advantage and feasibility of architecting things this way
    - The direction to go when building future features or refactoring
- At the moment the user API client calls the `User` model quite a lot, which is the wrong way round. But again, too much work to fix in this one pull request.

### CODE REVIEW

File: app/models/service.py [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+    @property
+    def trial_mode(self):
+        return self._dict['restricted']
+
+    def has_permission(self, permission):
+        return permission in self.permissions
+
+    @cached_property
+    def has_jobs(self):
+        # Can’t import at top-level because app isn’t yet initialised
```
Comment (leohemsted) [inline app/models/service.py]: comment no longer applies

File: app/templates/views/platform-admin/services.html [hunk 2]
```diff
@@ -47,7 +47,7 @@
 
         {% call row() %}
           {% call field(border=False) %}
-            <a href="{{ url_for('main.service_dashboard', service_id=service['id']) }}" class="browse-list-link">{{ service['name'] }}</a>
+            <a href="{{ url_for('main.service_dashboard', service_id=service.id) }}" class="browse-list-link">{{ service['name'] }}</a>
```
Comment (leohemsted) [inline app/templates/views/platform-admin/services.html]: should this have changed? It's on platform admin, so not handling `current_service`, but instead its own service obj

File: app/models/service.py [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+    def __bool__(self):
+        return self._dict != {}
+
+    def __getattr__(self, attr):
+        if attr in self.ALLOWED_PROPERTIES:
+            return self._dict[attr]
+        raise AttributeError('`{}` is not a service attribute'.format(attr))
+
+    def __getitem__(self, attr):
+        raise NotImplementedError(
```
[the remark this answers is not in the document]
Comment (quis) [reply app/models/service.py]: d3c48339

File: app/templates/views/platform-admin/services.html (hunk 2 above)
Comment (quis) [reply app/templates/views/platform-admin/services.html]: Jinja won’t care, but I’ve squashed it out for the sake of not changing things unnecessarily.

File: app/main/views/conversation.py [hunk 4]
```diff
@@ -44,13 +44,7 @@ def conversation_reply(
...
     notification_id,
 ):
-
-    templates = [
-        template
-        for template in service_api_client.get_service_templates(service_id)['data']
-        if template['template_type'] == 'sms'
-    ]
-
+    templates = current_service.templates_by_type('sms')
```
Comment (CrystalPea) [inline app/main/views/conversation.py]: neat ☺️

**Pull request merged**